### PR TITLE
Markdown code upgrade

### LINKED ISSUE #1829: Upgrade our markdown implementation to align with markdown-it-py==3.0.0
Validate/compare our code and plugins against the following to make relevant changes:
- [x] https://github.com/executablebooks/markdown-it-py/pull/267
Strict type annotation checking has been applied to the codebase in `markdown-it-py==3.0.0`.
Following which, the following mypy failures have surfaced and need to be corrected.

```
funnel/utils/markdown/mdit_plugins/heading_anchors_fix.py:45: error: Argument 3 to "after" of "Ruler" has incompatible type "Callable[[StateInline], None]"; expected "Callable[[StateCore], None]"  [arg-type]
funnel/utils/markdown/mdit_plugins/abbr.py:145: error: Argument 3 to "after" of "Ruler" has incompatible type "Callable[[StateInline], None]"; expected "Callable[[StateCore], None]"  [arg-type]
funnel/utils/markdown/mdit_plugins/toc.py:259: error: Argument 2 to "push" of "Ruler" has incompatible type "Callable[[StateInline], Any]"; expected "Callable[[StateCore], None]"  [arg-type]
funnel/utils/markdown/base.py:145: error: Argument 2 to "MarkdownIt" has incompatible type "dict[Literal['html', 'breaks', 'linkify', 'typographer'], bool]"; expected "Mapping[str, Any] | None"  [arg-type]
```
- [x] https://github.com/executablebooks/markdown-it-py/pull/283
- [x] https://github.com/executablebooks/markdown-it-py/pull/284

### PR DESCRIPTION
**Pull request by miteshashar**
Upgrades to markdown code to align with changes in markdown-it-py==3.0.0

Resolves #1829

### CODE REVIEW

File: funnel/utils/markdown/base.py [hunk 1]
```diff
@@ -89,7 +88,7 @@ class MarkdownConfig:
         'default', 'zero', 'commonmark', 'js-default', 'gfm-like'
     ] = 'commonmark'
     #: Updated options against the preset
-    options_update: Optional[Dict[OptionStrings, bool]] = None
+    options_update: Optional[Mapping] = None
```
Comment (jace) [inline funnel/utils/markdown/base.py:91]: This should be `Mapping[type, type]` or `dict` if not specifying types.

**Review by jace — COMMENTED** [review]
This needs a deps rebuild for 3.7. I'll add that.

**Comment by jace** [comment]
Just realised MDIT 3.0 drops Python 3.7, so this PR is blocked until we too drop 3.7.

File: funnel/utils/markdown/base.py (hunk 1 above)
Comment (miteshashar) [reply funnel/utils/markdown/base.py:91]: `Mapping` is equivalent to `Mapping[str, Any]`

**Pull request merged**